flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,533 @@
+    * @param out       the collector for outputting results
+    *
+    */
+  override def processElement2(
+    cRowValue: CRow,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    out: Collector[CRow]): Unit = {
+    val timeForRecord: Long = getTimeForRecord(ctx, cRowValue, false)
+    getCurrentOperatorTime(ctx)
+    processElement(
+      cRowValue,
+      timeForRecord,
+      ctx,
+      out,
+      rightOperatorTime,
+      leftOperatorTime,
+      leftTimerState,
+      rightCache,
+      leftCache,
+      false
+    )
+  }
+
+  /**
+    * Put a record from the input stream into the cache and iterate the opposite cache to
+    * output records meeting the join conditions. If there is no timer set for the OPPOSITE
+    * STREAM, register one.
+    */
+  private def processElement(
+    cRowValue: CRow,
+    timeForRecord: Long,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    out: Collector[CRow],
+    myWatermark: Long,
+    oppositeWatermark: Long,
+    oppositeTimeState: ValueState[Long],
+    recordListCache: MapState[Long, JList[Row]],
+    oppositeCache: MapState[Long, JList[Row]],
+    leftRecord: Boolean): Unit = {
+    if (relativeWindowSize > 0) {
+      //TODO Shall we consider adding a method for initialization with the context and collector?
+      cRowWrapper.out = out
+
+      val record = cRowValue.row
+
+      //TODO Only if the time of the record is greater than the watermark, can we continue.
+      if (timeForRecord >= myWatermark - allowedLateness) {
+        val oppositeLowerBound: Long =
+          if (leftRecord) timeForRecord - rightRelativeSize else timeForRecord - leftRelativeSize
+
+        val oppositeUpperBound: Long =
+          if (leftRecord) timeForRecord + leftRelativeSize else timeForRecord + rightRelativeSize
+
+        // Put the record into the cache for later use.
+        val recordList = if (recordListCache.contains(timeForRecord)) {
+          recordListCache.get(timeForRecord)
+        } else {
+          new util.ArrayList[Row]()
+        }
+        recordList.add(record)
+        recordListCache.put(timeForRecord, recordList)
+
+        // Register a timer on THE OTHER STREAM to remove records from the cache once they are
+        // expired.
+        if (oppositeTimeState.value == 0) {
+          registerCleanUpTimer(
+            ctx, timeForRecord, oppositeWatermark, oppositeTimeState, leftRecord, true)
+        }
+
+        // Join the record with records from the opposite stream.
+        val oppositeIterator = oppositeCache.iterator()
+        var oppositeEntry: Entry[Long, util.List[Row]] = null
+        var oppositeTime: Long = 0L;
+        while (oppositeIterator.hasNext) {
+          oppositeEntry = oppositeIterator.next
+          oppositeTime = oppositeEntry.getKey
+          if (oppositeTime < oppositeLowerBound - allowedLateness) {
+            //TODO Considering the data out-of-order, we should not remove records here.
```

Comment:
We can add a cleanup condition which is based on the watermark and not on the current records timestamps. If we make the cleanup timer a safety net with larger intervals, we can also clean up during joining. Should be more efficient since we go over the full MapState anyway.